File: packages/evals/tasks/bench/act/tab_handling.ts
import { z } from "zod";
import { defineBenchTask } from "../../../framework/defineTask.js";

export default defineBenchTask(
  { name: "tab_handling" },
  async ({ debugUrl, sessionUrl, stagehand, page, logger }) => {
    try {
      await page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/new-tab/");

      await stagehand.act("click the button to open the other page");

      const page2 = await stagehand.context.activePage();
      if (!page2) {
        throw new Error("No active page after opening the new tab");
      }
      const pages = await stagehand.context.pages();
      let page1: (typeof pages)[number] | undefined;
      for (const candidate of pages) {
        if (
          (await candidate.url()) ===
          "https://browserbase.github.io/stagehand-eval-sites/sites/new-tab/"
        ) {
          page1 = candidate;
          break;
        }
      }
      if (!page1) {
        throw new Error("Could not find the opener page after opening the new tab");
      }

      await stagehand.context.setActivePage(page1);
      const { data: extraction1 } = await stagehand.extract(
        "extract the entire page text",
        z.object({ extraction: z.string() }),
      );
      await stagehand.context.setActivePage(page2);
      const { data: extraction2 } = await stagehand.extract(
        "extract the entire page text",
        z.object({ extraction: z.string() }),
      );

      const extraction1Success = extraction1.extraction.includes("Welcome!");
      const extraction2Success = extraction2.extraction.includes("You’re on the other page");

      return {
        _success: extraction1Success && extraction2Success,
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    } catch (error) {
      return {
        _success: false,
        error: String(error),
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    } finally {
      await stagehand.close();
    }
  },
);

File: packages/sdk-ruby/evals/tasks/act/tab_handling.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/act/tab_handling.ts

Evals.define_task("tab_handling") do |t|
  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/new-tab/")

  t.stagehand.act("click the button to open the other page")

  # active_page polls until the popup registers (stagehand#2458), so
  # the page list is complete after this settles.
  t.stagehand.browser.context.active_page

  pages = t.stagehand.browser.context.pages
  page1 = pages[0]
  page2 = pages[1]

  # v3 used schemaless extract; v4 requires a schema.
  schema = {
    "type" => "object",
    "properties" => { "extraction" => { "type" => "string" } },
    "required" => ["extraction"],
    "additionalProperties" => false,
  }

  # extract all the text from the first page
  extraction1 = t.stagehand.extract("extract the entire page text", schema: schema, page: page1).data
  # extract all the text from the second page
  extraction2 = t.stagehand.extract("extract the entire page text", schema: schema, page: page2).data

  extraction1_success = extraction1["extraction"].include?("Welcome!")
  extraction2_success = extraction2["extraction"].include?("You’re on the other page")

  { _success: extraction1_success && extraction2_success }
end
